test/metric_value_benchmark/prometheus_test.go

```diff
@@ -0,0 +1,128 @@
+prometheus_test_gauge{include="yes"} 500
+# TYPE prometheus_test_summary summary
+prometheus_test_summary_sum{include="yes"} 200
+prometheus_test_summary_count{include="yes"} 50
+prometheus_test_summary{include="yes",quantile="0"} 0.1
+prometheus_test_summary{include="yes",quantile="0.5"} 0.25
+prometheus_test_summary{include="yes",quantile="1"} 5.5
+# TYPE prometheus_test_histogram histogram
+prometheus_test_histogram_sum{include="yes"} 300
+prometheus_test_histogram_count{include="yes"} 75
+prometheus_test_histogram_bucket{include="yes",le="0"} 1
+prometheus_test_histogram_bucket{include="yes",le="0.5"} 2
+prometheus_test_histogram_bucket{include="yes",le="2.5"} 3
+prometheus_test_histogram_bucket{include="yes",le="5"} 4
+prometheus_test_histogram_bucket{include="yes",le="+Inf"} 5
+`
+
+func (t *PrometheusTestRunner) validate() status.TestGroupResult {
+	metricsToFetch := t.getMeasuredMetrics()
+	testResults := make([]status.TestResult, len(metricsToFetch))
+	for i, metricName := range metricsToFetch {
+		testResults[i] = t.validatePrometheusMetric(metricName)
+	}
+
+	return status.TestGroupResult{
+		Name:        t.getTestName(),
+		TestResults: testResults,
+	}
+}
+
+func (t *PrometheusTestRunner) getTestName() string {
+	return "Prometheus"
+}
+
+func (t *PrometheusTestRunner) getAgentConfigFileName() string {
+	return "prometheus_config.json"
+}
+
+func (t *PrometheusTestRunner) getAgentRunDuration() time.Duration {
+	return time.Minute
+}
+
+func (t *PrometheusTestRunner) setupBeforeAgentRun() error {
+	startPrometheusCommands := []string{
+		fmt.Sprintf("cat <<EOF | sudo tee /tmp/prometheus_config.yaml\n%s\nEOF", prometheusConfig),
+		fmt.Sprintf("cat <<EOF | sudo tee /tmp/metrics\n%s\nEOF", prometheusMetrics),
+		"sudo python3 -m http.server 8101 --directory /tmp &> /dev/null &",
+	}
+
+	return common.RunCommands(startPrometheusCommands)
+}
+
+func (t *PrometheusTestRunner) setupAfterAgentRun() error {
+	return nil
+}
+
+func (t *PrometheusTestRunner) getMeasuredMetrics() []string {
+	return []string{
+		"prometheus_test_counter",
+		"prometheus_test_gauge",
+		"prometheus_test_summary_count",
+		"prometheus_test_summary_sum",
+		"prometheus_test_summary",
+	}
+}
+
+func (t *PrometheusTestRunner) validatePrometheusMetric(metricName string) status.TestResult {
+	testResult := status.TestResult{
+		Name:   metricName,
+		Status: status.FAILED,
+	}
+
+	fetcher, err := t.MetricFetcherFactory.GetMetricFetcher(metricName)
+	if err != nil {
+		log.Printf("%v", err)
+		return testResult
+	}
+
+	values, err := fetcher.Fetch(namespace, metricName, metric.AVERAGE)
+	if err != nil {
+		log.Printf("%v", err)
+		return testResult
+	}
+
+	if !isAllValuesGreaterThanOrEqualToZero(metricName, values) {
+		return testResult
+	}
+
+	// TODO: Range test with >0 and <100
+	// TODO: Range test: which metric to get? api reference check. should I get average or test every single datapoint for 10 minutes? (and if 90%> of them are in range, we are good)
```

Comment:
Question: have you considered just using the max?

Reply:
This is just a placeholder TODO from other tests. I think we still need to understand how to derive these benchmarks more broadly.